cmd/tether/ops_linux.go

```diff
@@ -129,19 +129,43 @@ func (t *operations) Setup(sink tether.Config) error {
 	return nil
 }
 
+// invoke will invoke the closure returned from the tether netfilter prep and
+// block until complete. It handles both potential preparation errors and invocation
+// errors.
+// the 'task' specified is used to construct error messages with the specific operation
+// embedded
+func invoke(t *tether.BaseOperations, fn tether.UtilityFn, preperr error, task string) error {
+	if preperr != nil {
+		return fmt.Errorf("%s preparation failed : %s", task, preperr)
+	}
+
+	exitChan, err := t.LaunchUtility(fn)
+	if err != nil {
+		return fmt.Errorf("%s failed: %s", task, err)
+	}
+
+	exitCode := <-exitChan
+	if exitCode != 0 {
+		return fmt.Errorf("%s returned non-zero: %d", task, exitCode)
+	}
+
+	return nil
+}
+
 // SetupFirewall sets up firewall rules on the external scope only.  Any
 // portmaps are honored as are port exposes.
-func (t *operations) SetupFirewall(config *tether.ExecutorConfig) error {
-	// XXX It looks like we'd want to collect the errors here, but we
-	// can't.  Since this is running inside init (tether) and tether
-	// reaps all children, the os.exec package won't be able to collect
-	// the error code in time before the reaper does.  The exec package
-	// calls wait and attempts to collect its child, but the reaper will
-	// have raptured the pid before that.  So, best effort, just keep going.
-	if err := netfilter.Flush(context.Background(), ""); err != nil {
+func (t *operations) SetupFirewall(ctx context.Context, config *tether.ExecutorConfig) error {
+	return setupFirewall(ctx, &t.BaseOperations, config)
+}
+
+// setupFirewall is broken out from SetupFirewall so that it can be referenced from the test code
```

Comment:
I tried for days when hitting the same problem with `Apply` in lib/tether for the network config and did not find a better solution. I agree, and in general the entire test code supporting tether is convoluted and there's massive copy/paste between the `cmd/x_test` and `lib/x_test` files. A lot of it comes down to function being in _test files to exclude it from production build, but Go preventing the use of exported function from those across packages. This was at a time before the compiler did dead code elimination.

Given this is impacting master I'm not going to dive into that rat hole again now. It does need to be addressed however, and in a more pervasive/coherent manner.